include/hashing/probing.h

```diff
@@ -0,0 +1,122 @@
+/**
+ * @file probing.h
+ * @author Sean Massung
+ *
+ * All files in META are dual-licensed under the MIT and NCSA licenses. For more
+ * details, consult the file LICENSE.mit and LICENSE.ncsa in the root of the
+ * project.
+ */
+
+#ifndef META_HASHING_PROBING_H_
+#define META_HASHING_PROBING_H_
+
+#include <cstdint>
+
+namespace meta
+{
+namespace hashing
+{
+namespace probing
+{
+
+class linear
+{
+  public:
+    linear(uint64_t hash, uint64_t capacity) : hash_{hash}, capacity_{capacity}
+    {
+        hash_ %= capacity_;
+    }
+
+    /**
+     * @return the next index to probe in the table
+     */
+    uint64_t probe()
+    {
+        return hash_++ % capacity_;
+    }
+
+  private:
+    uint64_t hash_;
+    uint64_t capacity_;
+};
+
+class linear_nomod
+{
+  public:
+    linear_nomod(uint64_t hash, uint64_t capacity) : hash_{hash}, max_{capacity - 1}
+    {
+        hash_ %= capacity;
+    }
+
+    /**
+     * @return the next index to probe in the table
+     */
+    uint64_t probe()
+    {
+        hash_++;
+        if (hash_ > max_)
+          hash_ = 0;
+        return hash_;
+    }
+
+  private:
+    uint64_t hash_;
+    uint64_t max_;
+};
+
+class binary
+{
+  public:
+    binary(uint64_t hash, uint64_t capacity) : hash_{hash}, step_{0}, capacity_{capacity}
+    {
+        hash_ %= capacity;
+    }
+
+    /**
+     * @return the next index to probe in the table
+     */
+    uint64_t probe()
+    {
+        // discard hashes that fall off of the table
+        for (; (hash_ ^ step_) >= capacity_; ++step_)
+          ;
+        return hash_ ^ step_++;
+    }
+
+  private:
+    uint64_t hash_;
+    uint64_t step_;
+    uint64_t capacity_;
+};
+
+// http://stackoverflow.com/questions/2348187
+//     /moving-from-linear-probing-to-quadratic-probing-hash-collisons
+class quadratic
+{
+  public:
+    quadratic(uint64_t hash, uint64_t capacity)
+        : hash_{hash}, capacity_{capacity}, step_{0}
+    {
+        hash_ &= (capacity_ - 1);
+    }
+
+    /**
+     * @note This strategy only will work for power-of-2 capacities!
+     * @return the next index to probe in the table
+     */
+    uint64_t probe()
```

Comment:
Do you think it's worth adding the discarding loop like in `binary_probing` so this probing strategy is at least able to run on non-power-of-2 sizes?